pkg/acpi/rsdp_linux.go

```diff
@@ -0,0 +1,200 @@
+func (r *RSDP) RSDPAddr() uint64 {
+	return r.base
+}
+
+// SDTAddr returns a base address or the [RX]SDT.
+//
+// It will preferentially return the XSDT, but if that is
+// 0 it will return the RSDT address.
+func (r *RSDP) SDTAddr() uint64 {
+	b := uint64(binary.LittleEndian.Uint32(r.data[16:20]))
+	if b != 0 {
+		return b
+	}
+	return binary.LittleEndian.Uint64(r.data[24:32])
+}
+
+func readRSDP(base uint64) (*RSDP, error) {
+	r := &RSDP{}
+	r.base = base
+
+	dat := memio.ByteSlice(make([]byte, len(r.data)))
+	if err := memio.Read(int64(base), &dat); err != nil {
+		return nil, err
+	}
+	copy(r.data[:], dat)
+	return r, nil
+}
+
+// GetRSDPEFI finds the RSDP in the EFI System Table.
+func GetRSDPEFI() (*RSDP, error) {
+	file, err := os.Open("/sys/firmware/efi/systab")
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	const (
+		acpi20 = "ACPI20="
+		acpi   = "ACPI="
+	)
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		start := ""
+		if strings.HasPrefix(line, acpi20) {
+			start = strings.TrimPrefix(line, acpi20)
+		}
+		if strings.HasPrefix(line, acpi) {
+			start = strings.TrimPrefix(line, acpi)
+		}
+		if start == "" {
+			continue
+		}
+		base, err := strconv.ParseUint(start, 0, 64)
+		if err != nil {
+			continue
+		}
+		rsdp, err := readRSDP(base)
+		if err != nil {
+			continue
+		}
+		return rsdp, nil
+	}
+	if err := scanner.Err(); err != nil {
+		log.Printf("error while reading EFI systab: %v", err)
+	}
+	return nil, fmt.Errorf("invalid /sys/firmware/efi/systab file")
+}
+
+// GetRSDPEBDA finds the RSDP in the EBDA.
+func GetRSDPEBDA() (*RSDP, error) {
+	f, err := os.OpenFile("/dev/mem", os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	e, err := ebda.ReadEBDA(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return getRSDPMem(uint64(e.BaseOffset), uint64(e.BaseOffset+e.Length))
+}
+
+func getRSDPMem(start, end uint64) (*RSDP, error) {
+	for base := start; base < end; base += 16 {
+		var r memio.Uint64
+		if err := memio.Read(int64(base), &r); err != nil {
+			continue
+		}
+		if r != 0x2052545020445352 {
+			continue
+		}
+		rsdp, err := readRSDP(base)
+		if err != nil {
+			return nil, err
+		}
+		return rsdp, nil
+	}
+	return nil, fmt.Errorf("could not find ACPI RSDP via /dev/mem from %#08x to %#08x", start, end)
+}
+
+// GetRSDPMem is the option of last choice, it just grovels through
+// the e0000-ffff0 area, 16 bytes at a time, trying to find an RSDP.
+// These are well-known addresses for 20+ years.
+func GetRSDPMem() (*RSDP, error) {
+	return getRSDPMem(0xe0000, 0xffff0)
+}
+
+// You can change the getters if you wish for testing.
+var getters = []func() (*RSDP, error){GetRSDPEFI, GetRSDPEBDA, GetRSDPMem}
+
+// GetRSDP finds the RSDP pointer and struct in memory.
+//
+// It is able to use several methods, because there is no consistency
+// about how it is done.
+func GetRSDP() (*RSDP, error) {
```

Comment:
we already have all these in pkg/boot/acpi -- can you merge the packages?

Reply:
I did not want to do that in this PR, but rather in another, easily reverted PR.